src/anyp/Uri.cc

```diff
@@ -514,27 +515,37 @@ AnyP::Uri::absolute() const
  *        After copying it on in the first place! Would be less code to merge the two with a flag parameter.
  *        and never copy the query-string part in the first place
  */
-char *
+SBuf
 urlCanonicalCleanWithoutRequest(const SBuf &url, const HttpRequestMethod &method, const AnyP::UriScheme &scheme)
 {
-    LOCAL_ARRAY(char, buf, MAX_URL);
-
-    snprintf(buf, sizeof(buf), SQUIDSBUFPH, SQUIDSBUFPRINT(url));
-    buf[sizeof(buf)-1] = '\0';
-
     // URN, CONNECT method, and non-stripped URIs can go straight out
+    SBuf out = url;
     if (Config.onoff.strip_query_terms && !(method == Http::METHOD_CONNECT || scheme == AnyP::PROTO_URN)) {
-        // strip anything AFTER a question-mark
-        // leaving the '?' in place
-        if (auto t = strchr(buf, '?')) {
-            *(++t) = '\0';
-        }
+        // strip anything after a path segment
+        // leaving the delimiter in place as a visual cue of truncation (if any)
+        static const CharacterSet pathDelims("path-delim", "?#");
+        const auto pathEnd = url.findFirstOf(pathDelims);
+        if (pathEnd != SBuf::npos)
+            out = url.substr(0, pathEnd+1);
     }
 
-    if (stringHasCntl(buf))
-        xstrncpy(buf, rfc1738_escape_unescaped(buf), MAX_URL);
+    // check for unescaped control characters
+    // XXX: emulates old stringHasCtl() which matched true
+    //      only for the UTF-8 characters 0x80-09f.
+    // TODO: RFC 1738, 5234 and 7230 require the full UTF-8
+    //      range (0x80-0xFF) be encoded along with
+    //      the ASCII control characters 0x01-0x1F, 0x7f
+    //      and NUL (0x00)
+    static const CharacterSet partialUtf8 = CharacterSet("utf8-partial", 0x80, 0x9f);
+    static const CharacterSet ctls = (partialUtf8 + CharacterSet::CTL)
+                                     .add('\0')
+                                     .rename("controls");
+    if (out.findFirstOf(ctls) != SBuf::npos) {
+        SBuf tmp = out;
+        out = SBuf(rfc1738_escape_unescaped(tmp.c_str()));
```

Comment:
True, that is a bug being preserved from the old code. I have added an XXX to mark it for a future fix. I believe it is not a major problem here because NUL detection is done specially in the message parser. It should still be fixed, for now this is preserving bug-compatibility.

Reply:
There are many bugs on this code path, most of them are indeed old. The _new_ bug this PR introduces is not inside the old rfc1738_escape_unescaped(). It is the added c_str() call. The new bug is masked by the old ones, and will disappear if the code is properly refactored in the future, but please add a second XXX to clearly mark the new bug so that whoever finishes this work or triages related problems is aware of this specific problem. For example:

```
// XXX: .c_str() below truncates strings containing NUL character
```

If you prefer just one XXX, then remove the XXX you added because it is vague and somewhat misleading -- rfc1738_escape_unescaped() itself does not really truncate NUL-containing c-strings because there is no such thing as a NUL-containing c-string (the problem is with the rfc1738_escape_unescaped() interface that should not be using c-strings). I am OK with two XXXs if you prefer that.

----

FWIW, the old urlCanonicalCleanWithoutRequest() code did _not_ call rfc1738_escape_unescaped() for strings containing NUL if there were no other special characters in the URL. Your code does that. Due to other old truncation bugs, the end result is the same AFAICT, but you are not just preserving the old rfc1738_escape_unescaped() bug here -- you are changing the kind of cases where that old bug is tickled.

IMO, this PR should adjust rfc1738_escape_unescaped() to properly solve all these related problems instead of adding new problems in order to preserve the old ones, but I do not insist on that bigger change at this time because I am still trying to understand what this PR scope and effects are.

Reply:
This is added in 3effc15f5

Reply:
> ... please add a second XXX to ... 
> If you prefer just one XXX, then remove the XXX you added ...

This change request is still unaddressed AFAICT. When you decide how to address it, please note the other old comment above:

> IMO, this PR should adjust rfc1738_escape_unescaped() to properly solve all these related problems instead of adding new problems in order to preserve the old ones, but I do not insist on that bigger change at this time

Reply:
This is resolved by 198201e